internal/server/data/access_key.go

```diff
@@ -112,21 +129,33 @@ func GetAccessKey(tx GormTxn, selectors ...SelectorFunc) (*models.AccessKey, err
 	return result, nil
 }
 
-func DeleteAccessKey(db GormTxn, id uid.ID) error {
-	return delete[models.AccessKey](db, id)
+type DeleteAccessKeysOptions struct {
+	// ByID instructs DeleteAccessKeys to delete the key with this ID.
+	ByID uid.ID
+	// ByUserID instructs DeleteAccessKeys to delete keys issued for this user.
+	ByUserID uid.ID
+	// ByProviderID instructs DeleteAccessKeys to delete keys issued by this
+	// provider.
+	ByProviderID uid.ID
 }
 
-func DeleteAccessKeys(db GormTxn, selectors ...SelectorFunc) error {
-	toDelete, err := list[models.AccessKey](db, nil, selectors...)
-	if err != nil {
-		return err
-	}
-
-	ids := slice.Map[models.AccessKey, uid.ID](toDelete, func(k models.AccessKey) uid.ID {
-		return k.ID
-	})
-
-	return deleteAll[models.AccessKey](db, ByIDs(ids))
+func DeleteAccessKeys(tx WriteTxn, opts DeleteAccessKeysOptions) error {
+	query := Query("UPDATE access_keys")
+	query.B("SET deleted_at = ? WHERE", time.Now())
+	switch {
+	case opts.ByID != 0:
```

Comment:
That works, but it's a slightly different pattern. I don't expect to see that pattern in write operations (insert, update, delete). I believe generally write operations only use one identifier.

Read operations, especially `list`, will very commonly use multiple options. You can see an example of that here from a much earlier draft:  https://github.com/infrahq/infra/pull/2657/files#diff-4d74142dc90ac5e0880f7655a7c285b71a1c3b1b03569066cfeb460a805a7038R94-R125

Instead of a `switch/case` we use `if` conditions. Unlike that early draft, I think we'll use `query.B` and explicitly add the `ADD` or `OR` between the clauses. I'll do that next for the access key get and list queries.